### PR TITLE
feat: expand OSC terminal integrations

### PR DESCRIPTION
**Pull request by depoll**
## Summary

Completes the practical OSC compatibility work on top of the original shell-metadata/image implementation:

- **Shell metadata:** OSC 7, 9;9, 133/633, and iTerm2 CurrentDir/RemoteHost working-directory support
- **Semantic command marks:** tracks OSC 133/633 command boundaries and iTerm2 `SetMark` with a bounded **Previous Command** terminal action
- **Dynamic colors:** applies OSC 4 setters across all xterm indexes 0–255 plus OSC 10/11/12/17 setters per session; OSC 104/110/111/112/117 restore the configured theme, and queries report the same effective palette the renderer uses
- **Kitty notifications:** OSC 99 capability negotiation; multipart title/body parsing; stable update/close identities; distinct unidentified notifications; bounded `p=alive` state; activation + close reports; `c=1` untracked-close responses; `u=0|1|2` urgency; `s=system|silent` sound; and bounded `w=` auto-expiration
- **iTerm2 controls:** `ReportCellSize` responses and `RequestAttention=yes|once|fireworks|no` through the native notification lifecycle
- **Inline images:** bounded iTerm2 OSC 1337 rendering through the existing Kitty graphics path, including cells, percentages, pixel dimensions, aspect-ratio behavior, and `doNotMoveCursor`

## Native Android evidence

The temporary live-probe entrypoint was not committed. It drove the production OSC parsers, terminal OSC callback, image renderer, command-mark tracker, and native notification service in a private-flavor Android build.

### Live compatibility probe

Native Android OSC compatibility probe

Visible checks cover:

- OSC 4/10/11/12 palette behavior and resets
- OSC 133/633 semantic marks and previous-command tracking
- Kitty OSC 99 lifecycle/query/report support
- iTerm2 cell-size reporting and pixel-sized inline image rendering
- iTerm2 attention request and notification close parsing

### Real OSC 1337 attention notification

Native Android OSC 1337 attention notification

The second capture is the Android notification shade after invoking the production `RequestAttention=once` path.

## Expanded native feature gallery

The gallery entrypoint was temporary and is not committed. Every state below is fed by the production OSC parser, session model, xterm renderer, or native notification service.

### Supported-feature overview

OSC supported-feature overview

### OSC 4 palette indexes 0–255

OSC 4 256-color palette setters, renderer, query, and reset

### Shell metadata, command marks, hyperlinks, and clipboard policy

OSC shell integration and command marks

### Kitty capability and alive negotiation

Kitty OSC 99 capability, alive, urgency, sound, and expiration

### iTerm2 inline graphics and cell-size response

iTerm2 OSC 1337 inline graphics and ReportCellSize

### Native urgency, sound, and silent delivery

Android notification shade showing critical, normal, and silent OSC notifications

## Post-review edge-case evidence

After an independent Opus / Sol / Grok council and the inline Copilot review, the temporary native gallery was expanded to exercise the corrected production paths. It remains uncommitted.

### Mixed colors, Cwd delimiter, Kitty identity/expiry/focus/close, and silent routing

Reviewed OSC protocol edge cases

### Aspect-preserving iTerm2 width×height containment

iTerm2 image fitted inside an 8x3 cell box

### Previous Command action and repeated jumps

Previous Command menu showing three semantic marks

Two consecutive Previous Command jumps reaching distinct rows

### Same-ID native update and real `p=close`

Android shade showing one updated same-ID Kitty notification

Production parser and service confirming p=close removed the native record

### Android channel-level silence

Android channel settings showing Sound None for the critical silent channel

## Follow-up review hardening

- iTerm2 attention uses a namespaced local identity, so it cannot enter Kitty `p=alive` state or alias a same-text Kitty ID.
- Unidentified Kitty native identities remain monotonic across shell restarts on one SSH connection.
- Oversized pixel dimension integers are parsed fallibly instead of escaping the OSC callback.
- Native notification presentation uses a bounded 64-entry per-connection queue that coalesces superseded same-identity updates/closes while preserving retained wire order.
- Disconnecting drops queued presentation work and clears a late in-flight native record instead of displaying stale notifications.
- Timed native notifications are cleared and removed from Kitty alive state on timeout, disconnect, provider disposal, and 256-record cap eviction.
- Preference-disabled updates leave an already-visible timed notification's cleanup state intact.
- Native show/clear exceptions release their current generation state before serialized processing continues.

## Safety and intentional limits

- Caps decoded OSC 1337 image payloads at 16 MiB and reuses existing decoded-image memory/dimension limits
- Does not retain OSC 633 command-line payloads or log paths, hosts, image names, or notification content
- Keeps remote file downloads unsupported
- Does not honor arbitrary focus stealing; attention requests use user-visible native notifications instead
- Bounds command anchors, pending Kitty messages, active identifiers, notification fields, expiration duration, and timer count
- Accepts only platform system/silent sounds and portable urgency levels; it does not accept arbitrary remote sound files, icons, or action buttons

## Validation

- `flutter analyze` — no issues
- `flutter test` — **3,112 tests passed**
- Focused OSC/theme/native/session tests — **290 passed**
- Terminal screen widget suite — **155 passed**
- Native Android live probe plus notification-shade capture
- Private Android APK build with JDK 17 — passed

### CODE REVIEW

File: lib/domain/services/terminal_iterm2_image.dart [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+/// Returns `true` for every `File=` command, including unsupported downloads or
+/// malformed payloads, so the caller can consume the command without logging
+/// user-controlled file metadata. Only `inline=1` payloads are decoded; remote
+/// file downloads are deliberately not supported.
+bool handleIterm2InlineImageOsc(Terminal terminal, List<String> args) {
+  if (args.isEmpty || !args.first.startsWith('File=')) {
+    return false;
+  }
+
+  final rawCommand = args.join(';');
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/terminal_iterm2_image.dart:46]: Fixed in c060832c: OSC 1337 now preflights both aggregate argument length and an independent 8 KiB metadata limit before joining. Added an aggregate-arguments regression.

File: lib/domain/services/terminal_iterm2_image.dart [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    graphicsArgs['r'] = '$rows';
+  }
+  if (metadata['doNotMoveCursor'] == '1') {
+    graphicsArgs['C'] = '1';
+  }
+
+  terminal
+    ..graphicsCommandStart(graphicsArgs)
+    ..graphicsDataChunk(imageBytes)
+    ..graphicsCommandEnd();
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/terminal_iterm2_image.dart]: Fixed in c060832c: vendored xterm now exposes atomic complete-command submission that refuses interleaving while a Kitty multipart command is active. OSC 1337 uses it, with a regression proving the pending Kitty image is not corrupted or finalized.

File: lib/domain/services/ssh_service.dart [hunk 3]
```diff
@@ -8046,20 +8185,27 @@ class ActiveSessionsNotifier extends Notifier<Map<int, SshConnectionState>> {
...
   ) async {
     if (!ref.mounted) return;
+    final notificationService = ref.read(localNotificationServiceProvider);
+    final notificationId = buildTerminalNotificationId(
+      session.connectionId,
+      identifier: request.identifier,
+    );
+    if (request.action == TerminalNotificationAction.close) {
+      await notificationService.clearTerminalNotification(notificationId);
+      return;
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/ssh_service.dart:8448]: Fixed in c060832c: terminal notification presentation is serialized per connection, preserving create, update, and close wire order. A delayed-native-show test proves p=close cannot overtake create.

File: lib/domain/services/ssh_service.dart [hunk 4]
```diff
@@ -1114,6 +1115,80 @@ Uri? parseTerminalWorkingDirectoryUri(List<String> args) {
...
+  String? value;
+  if (code == '9' && args.firstOrNull?.trim() == '9' && args.length > 1) {
+    value = args.skip(1).join(';');
+  } else if (code == '633' && args.firstOrNull == 'P') {
+    final propertyIndex = args.indexWhere((arg) => arg.startsWith('Cwd='));
+    if (propertyIndex >= 0) {
+      value = [
+        args[propertyIndex].substring('Cwd='.length),
+        ...args.skip(propertyIndex + 1),
+      ].join(';');
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/ssh_service.dart]: Fixed in c060832c: OSC 633 P now reads only the Cwd property value. Added coverage for P;Cwd=/home/demo/project; retaining the path without a trailing semicolon.

File: lib/domain/services/terminal_notification.dart [hunk 5]
```diff
@@ -91,9 +122,16 @@ class TerminalNotificationParser {
...
     final metadata = _parseMetadata(args.isNotEmpty ? args.first : '');
-    final id = metadata['i'] ?? '';
+    final id = _sanitizeIdentifier(metadata['i'] ?? '');
     final action = metadata['a'];
-    // Only the default (create/update) action produces a notification.
+    if (action == 'close') {
+      _pending.remove(id);
+      return id.isEmpty
+          ? null
+          : TerminalNotificationRequest.close(identifier: id);
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/terminal_notification.dart]: Addressed: production parses real Kitty p=close; a= is limited to focus and report behavior, including focus, -focus, report, and -report. Capability and parser tests use the real wire forms.

File: test/domain/services/terminal_notification_test.dart [hunk 6]
```diff
@@ -93,8 +93,22 @@ void main() {
       expect(req!.body, 'Encoded title');
     });
 
-    test('a close action does not emit a notification', () {
-      expect(parser.handleOsc('99', ['i=3:a=close']), isNull);
+    test('create/update and close actions preserve a bounded identifier', () {
+      final created = parser.handleOsc('99', ['i=build-3:a=create', 'Started']);
```
[the remark this answers is not in the document]
Comment (depoll) [reply test/domain/services/terminal_notification_test.dart]: Addressed: parser tests now create or update by sending no action with the same i=, and close with p=close. No a=create, a=update, or a=close expectations remain.

File: test/domain/services/ssh_session_terminal_notification_test.dart [hunk 7]
```diff
@@ -107,8 +200,22 @@ void main() {
...
     expect(request.body, 'All green');
+    expect(request.identifier, '1');
   });
 
+  test(
+    'routes Kitty close actions to the notification lifecycle stream',
+    () async {
+      final session = _session();
+      final next = session.terminalNotifications.first;
+      session.debugHandlePrivateOsc('99', const ['i=build:a=close']);
```
[the remark this answers is not in the document]
Comment (depoll) [reply test/domain/services/ssh_session_terminal_notification_test.dart]: Addressed: the session-level route test uses Kitty p=close, and the delayed presentation integration test also drives i=ordered:p=close through SshSession and ActiveSessionsNotifier.

File: lib/domain/services/ssh_service.dart [hunk 8]
```diff
@@ -4144,56 +4277,102 @@ class SshSession {
       return;
     }
 
+    final colorMutation = _terminalColorOverrides.handle(code, args);
+    if (colorMutation.handled) {
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/ssh_service.dart]: Fixed in c060832c: mutations are applied first, the effective theme is recomputed, then all query pairs are answered. Added an end-to-end OSC 4;1;#123456;1;? test asserting the new theme and post-mutation response.

File: lib/domain/services/ssh_service.dart [hunk 9]
```diff
@@ -8046,20 +8344,75 @@ class ActiveSessionsNotifier extends Notifier<Map<int, SshConnectionState>> {
...
+    final notificationId = buildTerminalNotificationId(
+      session.connectionId,
+      identifier: request.platformIdentifier,
+    );
+    final expiryKey = (
+      connectionId: session.connectionId,
+      notificationId: notificationId,
+    );
+    final generation = (_terminalNotificationGenerations[expiryKey] ?? 0) + 1;
+    _terminalNotificationGenerations[expiryKey] = generation;
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/ssh_service.dart]: Fixed in c060832c: presentation is serialized and generations now use a never-reused monotonic token. Exit paths compare before removal, so stale work cannot remove a superseding token.

File: lib/domain/services/ssh_service.dart [hunk 10]
```diff
@@ -8046,20 +8344,75 @@ class ActiveSessionsNotifier extends Notifier<Map<int, SshConnectionState>> {
...
+      title: title,
+      body: request.body,
+      urgency: request.urgency,
+      sound: request.sound,
+      timeout: request.timeout,
+      payload: TerminalNotificationPayload(
+        hostId: session.hostId,
+        connectionId: session.connectionId,
+        notificationIdentifier: request.identifier,
+        reportsActivation: request.reportsActivation,
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/ssh_service.dart]: Fixed in c060832c: terminal notification payload v3 carries the exact native notification ID separately from Kitty i=. Tap cleanup uses that ID. Added a no-i notification test proving its real expiry timer is removed.

File: lib/presentation/screens/terminal_screen.dart [hunk 11]
```diff
@@ -13818,6 +13830,36 @@ class _TerminalScreenState extends ConsumerState<TerminalScreen>
...
+    }
+    final lineHeight = viewState.renderTerminal.lineHeight;
+    if (!lineHeight.isFinite || lineHeight <= 0) return;
+    final position = _terminalScrollController.position;
+    final currentTopRow = (position.pixels / lineHeight).floor();
+    final beforeRow = position.pixels >= position.maxScrollExtent - 1
+        ? _terminal.buffer.height
+        : currentTopRow;
+    final targetRow = session.terminalCommandMarkTracker.previousMarkRow(
+      beforeRow,
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/presentation/screens/terminal_screen.dart]: Fixed in c060832c: Previous Command keeps a per-session and mark-count navigation cursor independent of clamped viewport offset and resets on independent scrolling, output, or new marks. Widget coverage invokes it twice and verifies the second jump reaches an older offset.

File: lib/domain/services/terminal_iterm2_image.dart [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    cellPixelExtent: cellPixelHeight,
+  );
+  final preservesAspectRatio = metadata['preserveAspectRatio'] != '0';
+  if (columns != null) {
+    graphicsArgs['c'] = '$columns';
+  }
+  // A single constrained dimension lets the renderer derive the other from
+  // the decoded image. When distortion is explicitly allowed, honor both.
+  if (rows != null && (columns == null || !preservesAspectRatio)) {
+    graphicsArgs['r'] = '$rows';
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/terminal_iterm2_image.dart]: Fixed in c060832c: encoded image dimensions and terminal cell aspect contain preserved images inside both requested dimensions; stretching remains only for preserveAspectRatio=0. Regression verifies a 1:1 image fits within a 20 by 4 box as 8 by 4, and native evidence shows 8 by 2 inside an 8 by 3 box.

File: lib/domain/services/local_notification_service.dart [hunk 13]
```diff
@@ -212,12 +310,25 @@ class LocalNotificationService {
...
   );
 
-  static const _terminalNotificationChannel = AndroidNotificationChannel(
-    terminalNotificationChannelId,
-    'Terminal notifications',
-    description: 'Notifications sent by the remote shell.',
-    importance: Importance.high,
-  );
+  static const _terminalNotificationChannels = <AndroidNotificationChannel>[
+    AndroidNotificationChannel(
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/local_notification_service.dart:352]: Fixed in c060832c: each urgency now has versioned system-sound and channel-level silent IDs. Silent channels use playSound false, and notification details also set silent true. Tests inspect channel construction; the PR includes Android Settings evidence showing Sound None.

**Comment by depoll** [comment]
Completed the requested Opus / Sol / Grok council review and addressed all 13 inline Copilot comments in `c060832c`, then merged the true latest `origin/main` in `269e917d`.

Additional council findings were also fixed: collision-free printable Kitty identifiers, Base64 pre-decode bounds, consecutive OSC 10/11/12 roles, `w=0`/`w=-1` never-expire semantics, `focus`/`-focus`, independent activation/close reporting, and xterm bright-white index 15.

Validation: full `flutter analyze` clean, all **3,086 Flutter tests passed**, targeted race/edge tests passed, and the private Android APK built with JDK 17.

The PR body now includes seven additional native captures for reviewed edge cases, 8×3 aspect containment, the Previous Command menu, two distinct consecutive jumps, same-ID native update, real `p=close`, and Android channel settings showing **Sound: None**. The temporary gallery was removed and the worktree is clean. Every inline thread has a specific resolution reply.

File: lib/domain/services/terminal_iterm2_control.dart [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    case 'fireworks':
+      return const TerminalNotificationRequest(
+        body: 'The remote terminal requested your attention.',
+        identifier: 'iterm2-attention',
+        sound: TerminalNotificationSound.system,
+      );
+    case 'no':
+      return const TerminalNotificationRequest.close(
+        identifier: 'iterm2-attention',
+      );
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/terminal_iterm2_control.dart:45]: Fixed in ec4c772d: iTerm2 attention show and close now use localIdentifier. Platform identities are explicitly namespaced as Kitty versus local, so attention neither enters Kitty alive state nor aliases a real Kitty i=iterm2-attention record. The control regression verifies both properties.

File: lib/domain/services/terminal_notification.dart [hunk 15]
```diff
@@ -2,60 +2,228 @@ import 'dart:convert';
...
+    final validated = _validatedKittyIdentifier(identifier ?? '');
+    if (validated.isNotEmpty) _activeIdentifiers.remove(validated);
+  }
+
+  /// Clears multipart and active notification state for a closed shell.
+  void reset() {
+    _pending.clear();
+    _activeIdentifiers.clear();
+    _unidentifiedSequence = 0;
+  }
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/terminal_notification.dart:241]: Fixed in ec4c772d: TerminalNotificationParser.reset now clears shell-scoped multipart and active state without rewinding the connection-scoped unidentified sequence. A regression creates notifications before and after reset and proves all native identities remain distinct.

File: lib/domain/services/terminal_iterm2_image.dart [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+    return ((availableCells * value) / 100).ceil().clamp(1, availableCells);
+  }
+  final pixels = RegExp(r'^(\d+)px$').firstMatch(raw);
+  if (pixels != null &&
+      cellPixelExtent != null &&
+      cellPixelExtent.isFinite &&
+      cellPixelExtent > 0) {
+    final pixelValue = int.parse(pixels.group(1)!);
+    if (pixelValue <= 0) return null;
+    return (pixelValue / cellPixelExtent).ceil().clamp(1, availableCells);
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/terminal_iterm2_image.dart]: Fixed in ec4c772d: pixel dimensions now use int.tryParse and reject null or non-positive values. A regression feeds a 4,096-digit width in px form and verifies the OSC command is safely consumed without throwing.

File: lib/domain/services/ssh_service.dart [hunk 17]
```diff
@@ -8041,25 +8344,123 @@ class ActiveSessionsNotifier extends Notifier<Map<int, SshConnectionState>> {
...
+  void _queueTerminalNotification(
+    SshSession session,
+    TerminalNotificationRequest request,
+  ) {
+    final connectionId = session.connectionId;
+    final previous =
+        _terminalNotificationQueues[connectionId] ?? Future<void>.value();
+    late final Future<void> operation;
+    operation = previous
+        .then((_) => _showTerminalNotification(session, request))
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/ssh_service.dart]: Fixed in ec4c772d: the Future chain is replaced by one bounded per-connection drain with at most 64 retained requests. Pending operations with the same namespaced platform identity are removed before the newest update or close is appended, preserving the relative wire order of all retained identities. Regressions prove an interleaved other identity stays ordered, a superseded show becomes one close, and a 512-message uncoalesced flood retains exactly the configured bound.

**Comment by depoll** [comment]
Addressed the four new inline findings in `ec4c772d` and the three suppressed lifecycle findings from the same Copilot review in `8361a16b`.

Follow-up lifecycle hardening now also:
- stops queued operations after detach and clears an in-flight record that finishes late;
- clears timed native notifications on disconnect/provider disposal;
- clears and closes the oldest timed record when the 256-record cap is reached instead of silently dropping its timer.

Added focused regressions for a 512-message queue flood, same-ID coalescing with retained ordering, disconnect during an in-flight show, disconnect with a timed notification, and 257-record expiry eviction.

Validation: full `flutter analyze` clean, all **3,110 Flutter tests passed**, focused notifier tests passed, the direct local-notification suite passed, and the private Android APK built with JDK 17. The PR body is updated and all four new inline threads are resolved.

File: lib/domain/services/ssh_service.dart [hunk 18]
```diff
@@ -8041,25 +8354,139 @@ class ActiveSessionsNotifier extends Notifier<Map<int, SshConnectionState>> {
...
+    final timeout = request.timeout;
+    if (timeout == null) {
+      _removeTerminalNotificationGeneration(expiryKey, generation);
+      return;
+    }
+    const maxExpiryTimers = 256;
+    if (_terminalNotificationExpiryTimers.length >= maxExpiryTimers) {
+      final oldest = _terminalNotificationExpiryTimers.keys.first;
+      _terminalNotificationExpiryTimers.remove(oldest)?.cancel();
+      _terminalNotificationGenerations.remove(oldest);
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/ssh_service.dart]: Addressed in 8361a16b before this thread arrived: each retained expiry now carries its session, protocol identifier, and notification service. Cap eviction runs the same cleanup as timeout and detach, canceling the timer, removing generation state, closing the Kitty active identifier, and clearing the native notification. The 257-record regression verifies the oldest native ID is cleared.

File: lib/domain/services/ssh_service.dart [hunk 19]
```diff
@@ -8041,25 +8354,139 @@ class ActiveSessionsNotifier extends Notifier<Map<int, SshConnectionState>> {
...
+      connectionId: session.connectionId,
+      notificationId: notificationId,
+    );
+    final generation = ++_nextTerminalNotificationGeneration;
+    _terminalNotificationGenerations[expiryKey] = generation;
+    _terminalNotificationExpiryTimers.remove(expiryKey)?.cancel();
+    if (request.action == TerminalNotificationAction.close) {
+      await notificationService.clearTerminalNotification(notificationId);
+      _removeTerminalNotificationGeneration(expiryKey, generation);
+      return;
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/ssh_service.dart]: Fixed in 4a166de3: close uses try/finally to remove its current generation even when native clear throws, and native show removes its generation in catch before rethrowing to the queue error reporter. A controlled throwing-plugin regression verifies both paths leave zero generations.

File: lib/domain/services/ssh_service.dart [hunk 20]
```diff
@@ -8041,25 +8354,139 @@ class ActiveSessionsNotifier extends Notifier<Map<int, SshConnectionState>> {
...
+    }
+    if (!ref.read(terminalNotificationsNotifierProvider)) {
+      _removeTerminalNotificationGeneration(expiryKey, generation);
+      return;
```
[the remark this answers is not in the document]
Comment (depoll) [reply lib/domain/services/ssh_service.dart:8448]: Fixed in 4a166de3: show requests check the current terminal-notification preference before assigning a generation or canceling an existing expiry. Close still runs while disabled. A regression presents a timed record, disables notifications, sends a same-ID update, and verifies both its timer and native call history remain unchanged.
